---
title: "Identifying Active and Alternative Promoters from RNA-Seq data with proActiv"
author:
  - name: Joseph Lee, Deniz Demircioğlu, Jonathan Göke 
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Identifying Active and Alternative Promoters from RNA-Seq data with proActiv}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Summary

Most human genes have multiple promoters that control the expression of distinct isoforms. The use of these alternative promoters enables the regulation of isoform expression pre-transcriptionally. Alternative promoters have been found to be important in a wide number of cell types and diseases.

proActiv is a method that enables the analysis of promoters from RNA-Seq data. proActiv uses aligned reads as input, and then generates counts and normalized promoter activity estimates for each annotated promoter. These estimates can then be used to identify which promoter is active, which promoter is inactive, and which promoters change their activity across conditions.

Here we present a quick start guide to using proActiv, and a detailed workflow for identifying active promoters and alternative promoters across 2 conditions.

If you use proActiv in your research, please cite: 

[Demircioğlu, Deniz, et al. "A Pan-cancer Transcriptome Analysis Reveals Pervasive Regulation through Alternative Promoters." *Cell* 178.6 (2019): 1465-1477.](https://www.cell.com/cell/fulltext/S0092-8674(19)30906-7)

## Contents

- [Quick Start: Quantifying promoter activity with proActiv]
- [A complete workflow to identify alternative promoter usage]
  - [Preparing input data]
  - [Preparing promoter annotations]
  - [Running proActiv]
  - [Alternative promoter usage]
- [Analysis and visualization of alternative promoter usage]
  - [Alternative promoter usage]
  - [Promoter category proportion]
  - [Major/minor promoters by position]
  - [Major promoter activity and gene expression]
  - [t-SNE]
- [Getting help]
- [Citing proActiv]
- [Session information]

## Quick Start: Quantifying promoter activity with proActiv

proActiv estimates promoter activity from RNA-Seq data. Promoter activity is defined as the total amount of transcription initiated at each promoter. proActiv takes as input either BAM files or junction files (TopHat2 or STAR), and a promoter annotation object of the relevant genome. An optional argument `condition` can be supplied, describing the condition corresponding to each input file. Here we demonstrate proActiv with STAR junction files (Human genome GRCh38 Gencode v34) as input:

```{r QuickStart, eval=TRUE, message=FALSE, results='hide'}
library(proActiv)

## List of STAR junction files as input
files <- list.files(system.file('extdata/vignette/junctions', 
                                package = 'proActiv'), full.names = TRUE)
## Vector describing experimental condition
condition <- rep(c('A549','HepG2'), each=3)
## Promoter annotation for human genome GENCODE v34
promoterAnnotation <- promoterAnnotation.gencode.v34

result <- proActiv(files = files, 
                   promoterAnnotation = promoterAnnotation,
                   condition = condition)
```

`result` is a summarizedExperiment object which can be accessed as follows:

 - `assays(results)` returns raw/normalized promoter counts and absolute/relative promoter activity  
 - `metadata(results)` returns gene expression data  
 - `rowData(results)` returns promoter metadata and summarized absolute promoter activity by conditions

## A complete workflow to identify alternative promoter usage

Here we present a complete step-by-step workflow for analysing promoter activity with proActiv and for identifying alternative promoter usage across samples from different conditions. We will compare samples from 2 different cell lines (A549 and HepG2) to identify alternative promoters.

### Preparing input data

proActiv uses RNA-Seq data to quantify promoter activity. Users have the option of using as input either BAM files, or tab-delimited junction files that are generated when TopHat2 or STAR is used for read alignment.

Below, we demonstrate running `proActiv` with input STAR junction files. This data is taken from the [SGNEx project](https://github.com/GoekeLab/sg-nex-data). The reference genome used for alignment is Gencode v34 (GRCh38). These files can be found in ‘extdata/vignette/junctions’:

- extdata/vignette/junctions/SGNEx_A549_Illumina_replicate1-run1.junctions.gz
- extdata/vignette/junctions/SGNEx_A549_Illumina_replicate3-run1.junctions.gz 
- extdata/vignette/junctions/SGNEx_A549_Illumina_replicate5-run1.junctions.gz
- extdata/vignette/junctions/SGNEx_HepG2_Illumina_replicate2-run1.junctions.gz 
- extdata/vignette/junctions/SGNEx_HepG2_Illumina_replicate4-run1.junctions.gz 
- extdata/vignette/junctions/SGNEx_HepG2_Illumina_replicate5-run1.junctions.gz

```{r ListFiles, eval=FALSE}
files <- list.files(system.file('extdata/vignette/junctions', 
                                package = 'proActiv'), full.names = TRUE)
```

### Preparing promoter annotations

In order to quantify promoter activity, proActiv uses a set of promoters based on genome annotations. proActiv allows the creation of a promoter annotation object for any genome from a TxDb object or from a GTF file with the `preparePromoterAnnotation` function. Users have the option to either pass the file path of the GTF/GFF or TxDb to be used, or use the TxDb object directly as input. proActiv includes pre-calculated promoter annotations for human and mouse genomes:

- GENCODE Release 19 / GRCh37 / hg19 : `promoterAnnotation.gencode.v19` 
- GENCODE Release 34 / GRCh38 / hg38 : `promoterAnnotation.gencode.v34`
- GENCODE Release M1 / NCBIM37 / mm9 : `promoterAnnotation.gencode.vM1`
- GENCODE Release M25 / GRCm38.p6 / mm10 : `promoterAnnotation.gencode.vM25`  

GTF files can be downloaded from the [GENCODE](https://www.gencodegenes.org) page.

Here, we demonstrate creating the promoter annotation for the Human genome (GENCODE v34) with both GTF and TxDb. To keep the run-time of this vignette small, we use a subset of the GTF/TxDb which includes only chromosome 22 annotations.

```{r CreateAnnotation, eval=FALSE}
## From GTF file path
gtf.file <- system.file('extdata/vignette/annotation/gencode.v34.annotation.chr22.gtf.gz', 
                        package = 'proActiv')
promoterAnnotation.gencode.v34.chr22 <- preparePromoterAnnotation(file = gtf.file,
                                                                  species = 'Homo_sapiens')
## From TxDb object
txdb.file <- system.file('extdata/vignette/annotation/gencode.v34.annotation.chr22.sqlite', 
                         package = 'proActiv')
txdb <- loadDb(txdb.file)
promoterAnnotation.gencode.v34.chr22 <- preparePromoterAnnotation(txdb = txdb, 
                                                                  species = 'Homo_sapiens')
```

The `PromoterAnnotation` object has 3 slots:
 
- `intronRanges` : Intron ranges, giving the corresponding transcripts of each intron
- `promoterIdMapping` : An ID mapping between transcripts, promoter IDs and gene IDs  
- `promoterCoordinates` : Promoter coordinates (TSS) and internal promoter state, along with the 3' coordinate of the first exon 

### Running proActiv

Once promoters in the genome are identified, proActiv estimates promoter activity at each annotated promoter. Here, we load pre-calculated promoter annotation for GENCODE Release 34. We also supply the experimental condition to `proActiv`. This information allows `proActiv` to summarize results across conditions.

```{r proActiv, eval=FALSE, message=FALSE, results='hide'}
promoterAnnotation <- promoterAnnotation.gencode.v34

condition <- rep(c('A549', 'HepG2'), each=3)

result <- proActiv(files = files,
                   promoterAnnotation = promoterAnnotation,
                   condition = condition)
```

`result` is a `SummarizedExperiment` object with assays as raw/normalized promoter counts, and absolute/relative promoter activity: 

```{r proActiv result, eval=TRUE}
show(result)
``` 

The `rowData` slot stores a promoter-gene ID mapping and promoter position (5' to 3') for each promoter by gene. Mean absolute promoter activity for each condition is also summarized here. Promoters are also categorized into three classes. Promoters with activity < 0.25 are classified as inactive, while the most active promoters of each gene are classified as major promoters. Promoters active at lower levels are classified as minor promoters. 

```{r Result rowData, eval=TRUE, echo=FALSE}
knitr::kable(head(rowData(result)))
```

The `metadata` slot provides gene expression data for each replicate per condition, and also summarizes mean expression across conditions.

For cleaner downstream analysis, one can remove single-exon transcripts for which promoter activity is not quantified. `result` can be filtered as such:

```{r filter result, eval=TRUE}
## Removes single-exon transcripts / promoters by eliminating promoter counts that are NA 
result <- result[complete.cases(assays(result)$promoterCounts),]
```

### Alternative promoter usage

We identify genes with similar expression levels across cell lines with alternate promoter usage using [DEXSeq](https://bioconductor.org/packages/release/bioc/html/DEXSeq.html) (Anders, Reyes, Huber 2012). While DEXSeq is originally intended for inferring differential exon usage in RNA-Seq data, DEXSeq can be similarly used with raw promoter counts as input to assess the statistical significance of alternative promoter usage across conditions. We follow the standard [DEXSeq workflow](https://bioconductor.org/packages/devel/bioc/vignettes/DEXSeq/inst/doc/DEXSeq.html#3_counting_reads). To keep the run-time of this vignette short, we focus only on annotated promoters from chromosome 1.      

```{r DEXSeq, eval=TRUE, message=FALSE, warning=FALSE}
library(DEXSeq)

## Subset to chromosome 1
result.chr <- result[rowData(result)$seqnames == 'chr1', ]
countData <- data.frame(assays(result.chr)$promoterCounts, rowData(result.chr))

## Call DEXSeq - promoter as feature, gene as group
dxd <- DEXSeqDataSet(countData = as.matrix(countData[,seq_len(length(condition))]),
                     sampleData = data.frame(colData(result.chr)),
                     design = formula(~ sample + exon + condition:exon),
                     featureID = as.factor(countData$promoterId),
                     groupID = as.factor(countData$geneId))
dxr1 <- DEXSeq(dxd)
``` 

`DEXSeq` returns `dxr1`, which provides the significance that a promoter is differentially used across conditions. The description of each column of `dxr1` can be found in the metadata columns. 

```{r DEXSeq column description, eval=TRUE}
mcols(dxr1)$description
```

Users can sort this result by the minimum adjusted p-value for all promoters belonging to a particular gene. This helps to better identify candidate genes where alternative promoter usage is present.    

```{r dxr1 wrangling, eval=TRUE}
## Arrange by minimum padj for each gene
dxr1 <- data.frame(dxr1[,1:10]) %>% 
  group_by(groupID) %>% 
  mutate(minp = min(padj)) %>%
  arrange(minp)
```

```{r dxr1, eval=TRUE, echo=FALSE, layout='l-body-outset'}
knitr::kable(head(data.frame(dxr1)))
``` 

## Analysis and visualization of alternative promoter usage

Here we offer several visualizations of the data returned by the workflow above.

### Alternative Promoter Usage

To visualize genes with alternative promoter usage across conditions, we call `plotPromoters` on the summarizedExperiment object `result` returned by proActiv. Here, we demonstrate `plotPromoters` by visualizing promoters of RAP1GAP (ENSG00000076864.19). RAP1GAP is the second most significant gene identified in `dxr1` as a candidate for alternative promoter usage across cell lines A549 and HepG2. `plotPromoters` takes in `result` and `gene`, a gene of interest. 

In order to build and plot a transcript model for the gene of interest, users may supply either a transcript database (`txdb`) or a list of Genomic Ranges giving the ranges of exons by transcripts to be plotted (`ranges`). If users choose to use a TxDb as input, we recommend that the TxDb used should be the same as the one used to prepare promoter annotations, as annotations from different sources may differ slightly. To keep the run-time of this vignette short, we use a TxDb generated from GENCODE v34 GTF subsetted to RAP1GAP. 

```{r VizAlternativePromoters txdb, eval=TRUE, message=FALSE, warning=FALSE, fig.align='center', fig.height=7, fig.width=7}
## RAP1GAP
gene <- 'ENSG00000076864.19'
txdb <- loadDb(system.file('extdata/vignette/annotations',
                           'gencode.v34.annotation.rap1gap.sqlite', 
                           package = 'proActiv'))
plotPromoters(result = result, gene = gene, txdb = txdb)
```

The same plot can be generated with a list of Genomic Ranges giving the exons by transcripts of RAP1GAP:

```{r VizAlternativePromoters ranges, eval=FALSE}
ranges <- readRDS(system.file('extdata/vignette/annotations',
                              'exonsBy.rap1gap.rds',
                              package = 'proActiv'))
plotPromoters(result = result, gene = gene, ranges = ranges)
```

Users can adjust the width of the promoter 'blocks' and 'arrows' in the plot with the numeric arguments `blk.width` and `arrow.width` respectively. `blk.width` defaults to 500 (bases), while `arrow.width` is internally calculated based on the range of the gene. Other parameters controlling the fill and border colour and size of labels are listed in code documentation. 

From the plot, it is clear that alternative promoter usage regulates the expression of RAP1GAP across the cell lines A549 and HepG2. 

### Promoter category proportion 

Here, we visualize the categorization of annotated promoters in the two cell lines. The proportions between the categories are similar across the two cell lines, with majority of the promoters being inactive. 

```{r VizPromoterCategories, eval=TRUE, fig.align='center', fig.height=5, fig.width=5, message=FALSE, warning=FALSE}
library(ggplot2)

rdata <- rowData(result)
## Create a long dataframe summarizing cell line and promoter class
pdata1 <- data.frame(cellLine = rep(c('A549', 'HepG2'), each = nrow(rdata)),
                       promoterClass = as.factor(c(rdata$A549.class, rdata$HepG2.class)))

ggplot(na.omit(pdata1)) +
  geom_bar(aes(x = cellLine, fill = promoterClass)) + 
  xlab('Cell Lines') + ylab('Count') +  labs(fill = 'Promoter Category') +
  ggtitle('Categorization of Promoters')
```

### Major/minor promoters by position

Analysis of major:minor promoter proportions against promoter position. The analysis is restricted to multi-promoter genes with at least one active promoter. Below, we generate the plot for cell line A549. In general, the major:minor promoter proportion decreases with increasing promoter position.

```{r VizMajorMinorPosition, eval=TRUE, fig.align='center', fig.height=5, fig.width=5, message=FALSE, warning=FALSE}
## Because many genes have many annotated promoters, we collapse promoters 
## from the 5th position and onward into one group for simplicity
pdata2 <- as_tibble(rdata) %>%
  mutate(promoterPosition = ifelse(promoterPosition > 5, 5, promoterPosition)) %>%
  filter(A549.class %in% c('Major', 'Minor'))

ggplot(pdata2) +
  geom_bar(aes(x = promoterPosition, fill = as.factor(A549.class)), position = 'fill') +
  xlab(expression(Promoter ~ Position ~ "5'" %->% "3'")) + ylab('Percentage') + 
  labs(fill = 'Promoter Category') + ggtitle('Major/Minor Promoter Proportion in A549') + 
  scale_y_continuous(breaks = seq(0,1, 0.25), labels = paste0(seq(0,100,25),'%')) +
  scale_x_continuous(breaks = seq(1,5), labels = c('1','2','3','4','>=5'))
```

### Major promoter activity and gene expression

Comparison of major promoter activity and gene expression, calculated by summing over all promoters. Single promoter genes lie on the diagonal. Multi-promoter genes lie to the right of the diagonal. Below, we generate the plot for cell line A549. This plot suggests that a single major promoter does not often fully explain gene expression, with minor promoters also contributing to gene expression.

```{r VizMajorGeneExp, eval=TRUE, fig.align='center', fig.height=5, fig.width=6, message=FALSE, warning=FALSE}
## Get active major promoters of A549
majorPromoter <- as_tibble(rdata) %>% group_by(geneId) %>% 
  mutate(promoterCount = n()) %>% filter(A549.class == 'Major') 
## Get gene expression corresponding to the genes identified above
geneExpression <- metadata(result)$geneExpression %>% 
  rownames_to_column(var = 'geneId') %>% 
  filter(geneId %in% majorPromoter$geneId)

pdata3 <- data.frame(proActiv = majorPromoter$A549.mean,
                     geneExp = geneExpression$A549.mean[match(majorPromoter$geneId, 
                                                              geneExpression$geneId)],
                     promoterCount = majorPromoter$promoterCount)

ggplot(pdata3, aes(x = geneExp, y = proActiv)) + 
  geom_point(aes(colour = promoterCount), shape = 1) +
  ggtitle('Major Promoter Activity vs. Gene Expression in A549') + 
  xlab('Average Gene Expression') + ylab('Average Major Promoter Activity') +
  labs(colour = 'Number of \n Annotated Promoters') +
  geom_abline(slope = 1, intercept = 0, colour = 'red', linetype = 'dashed')
```

### t-SNE 

We generate a t-SNE plot with all active promoters. Expectedly, replicates from each cell line cluster together. 

```{r VizTsne, eval=TRUE, fig.align='center', fig.height=5.2, fig.width=5.2}
library(Rtsne)

## Remove inactive promoters (sparse rows)
data <- assays(result)$absolutePromoterActivity %>% filter(rowSums(.) > 0)
data <- data.frame(t(data))
data$Sample <- as.factor(condition)

set.seed(42) # for reproducibility

tsne.out <- Rtsne(as.matrix(subset(data, select = -c(Sample))), perplexity = 1)
plot(x = tsne.out$Y[,1], y = tsne.out$Y[,2], bg = data$Sample, asp = 1,
     col = 'black', pch = 24, cex = 4,
     main = 't-SNE plot with promoters \n active in at least one sample',
     xlab = 'T-SNE1', ylab = 'T-SNE2',
     xlim = c(-180,180), ylim = c(-180,180))
legend('topright', inset = .02, title = 'Cell Lines',
       unique(condition), pch = c(24,24), pt.bg = 1:length(unique(condition)) , cex = 1.5, bty = 'n')
```

## Getting help

Questions and issues can be raised at the Bioconductor support site: https://support.bioconductor.org. Ensure your posts are tagged with `proActiv`. 

Alternatively, issues can be raised at the proActiv Github repository: https://github.com/GoekeLab/proActiv.

## Citing proActiv

If you use proActiv, please cite:

[Demircioğlu, Deniz, et al. "A Pan-cancer Transcriptome Analysis Reveals Pervasive Regulation through Alternative Promoters." *Cell* 178.6 (2019): 1465-1477.](https://www.cell.com/cell/fulltext/S0092-8674(19)30906-7)

## Session information

```{r SessionInfo, eval=TRUE, echo=FALSE}
sessionInfo()
```
