---
title: "Identifying Alternative Promoter Usage with proActiv"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Identifying Alternative Promoter Usage with proActiv}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Workflow

Here we present a workflow for analysing promoter activity with proActiv. Promoter activity is defined as the total amount of transcription initiated at each promoter. proActiv takes as input either BAM files or junction files generated by Tophat or STAR aligners, and infers promoter counts and activity. For this workflow, STAR junction files from RNA-seq of two cell lines (A549 and MCF7) with three replicates each are used.

First, we load the promoter annotation object and call `proActiv`.


```{r eval=TRUE, results = 'hide', message=FALSE}
library(proActiv)
library(SummarizedExperiment)
library(dplyr)

files <- list.files(system.file('extdata/vignette', 
                                package = 'proActiv'), full.names = TRUE)

promoterAnnotation <- proActiv::promoterAnnotation.gencode.v34

result <- proActiv::proActiv(files = files,
                   promoterAnnotation = promoterAnnotation)
```

`result` is a `SummarizedExperiment` object with assays comprising promoter counts and activity. Gene expression is stored as metadata and a promoter-to-gene map is stored as row data.

Next, we summarize promoter activity and gene expression across replicates for each group, A549 and MCF7.    

```{r eval=TRUE}
groups <- c('A549', 'MCF7')

geneExp <- metadata(result)$geneExpression %>%
  transmute(geneId = rownames(.),
            A549 = rowMeans(select(., contains(groups[1]))),
            MCF7 = rowMeans(select(., contains(groups[2]))))

mapper <- as_tibble(rowData(result))

proCoord <- as_tibble(promoterCoordinates(promoterAnnotation)) %>%
  bind_cols(geneId = mapper$geneId)

proActiv <- as_tibble(assays(result)$absolutePromoterActivity) %>% 
  transmute(A549 = rowMeans(select(., contains(groups[1]))),
            MCF7 = rowMeans(select(., contains(groups[2])))) %>%
  bind_cols(chr = proCoord$seqnames, mapper, .) %>%
  filter(chr == 'chr3')
```

For each cell line, `proActiv` contains the average absolute promoter activity across replicates: 

```{r eval=TRUE, echo=FALSE}
knitr::kable(head(proActiv))
```

### Promoter Position

Next, we extract the position of each promoter (5' to 3'), and append this information to `proActiv`:

```{r, eval=TRUE}
proCoord <- proCoord %>%
  group_by(geneId) %>%
  mutate(promoterPosition = ifelse(strand == '+', row_number(), rev(row_number()))) %>%
  mutate(promoterRank = ifelse(promoterPosition > 5, 5, promoterPosition))

proActiv$promoterPosition = proCoord$promoterPosition[match(proActiv$promoterId, proCoord$promoterId)]

proActiv$promoterRank = proCoord$promoterRank[match(proActiv$promoterId, proCoord$promoterId)]
```

### Promoter Category

Promoters are categorized into three categories. The most active promoter for each gene is categorized as a major promoter, while additional promoters that are active at lower levels are categorized as minor promoters. Promoters with promoter activity less than 0.25 are categorized as inactive promoters.

```{r eval=TRUE}
proActiv <- proActiv %>%
  group_by(geneId) %>%
  mutate(promoterCount = n()) %>%
  arrange(desc(A549), promoterPosition) %>%
  mutate(A549_class = cut(A549, breaks = c(-Inf, 0.25, Inf), labels = c('Inactive', 'Minor'), right = FALSE),
         A549_class = ifelse(row_number() == 1 & dplyr::first(A549_class) == 'Minor', 'Major', as.character(A549_class))) %>%
  arrange(desc(MCF7), promoterPosition) %>%
  mutate(MCF7_class = cut(MCF7, breaks = c(-Inf, 0.25, Inf), labels = c('Inactive', 'Minor'), right = FALSE),
         MCF7_class = ifelse(row_number() == 1 & dplyr::first(MCF7_class) == 'Minor', 'Major', as.character(MCF7_class))) %>%
  arrange(geneId, promoterPosition)
```

### Differential Promoter Usage with DEXSeq

We identify genes with similar expression levels across cell lines with alternate promoter usage using [DEXSeq](https://bioconductor.org/packages/release/bioc/html/DEXSeq.html) (Anders, Reyes, Huber 2012). While DEXSeq is originally intended for inferreing differential exon usage in RNA-Seq data, DEXSeq can be similarly used with raw promoter counts as input to assess the statistical significance of differential promoter usage across conditions. We follow the standard [DEXSeq workflow](https://bioconductor.org/packages/devel/bioc/vignettes/DEXSeq/inst/doc/DEXSeq.html#3_counting_reads).      

```{r, eval=TRUE, message=FALSE, warning=FALSE}
library(DEXSeq)

data <- bind_cols(chr = proCoord$seqnames, mapper, assays(result)$promoterCounts) %>% 
  filter(chr == 'chr3') %>% 
  na.omit 

countData <- as.matrix(data[,-c(1:3)])

sampleData <- data.frame(condition = as.factor(rep(groups, each=3)))

design <- formula( ~ sample + exon + condition:exon )

groupID <- as.factor(data$geneId)

featureID <- as.factor(data$promoterId)

dxd <- DEXSeqDataSet(countData, sampleData, design,
                     featureID, groupID)

dxd = estimateSizeFactors(dxd)

dxd = estimateDispersions(dxd)

dxd = testForDEU(dxd)

dxd = estimateExonFoldChanges( dxd, fitExpToVar="condition")

dxr1 = DEXSeqResults(dxd)
```

Because we are interested in genes whose expression levels differ minimally but utilise alternate promoters, we filter the results returned by DEXSeq. For simplicity, we restrict the analysis to genes with fewer than five promoters.  

```{r eval=TRUE}
diffPromoter <- as_tibble(dxr1) %>%
  dplyr::select(groupID, featureID, pvalue, padj) %>%
  mutate(A549_class = proActiv$A549_class[match(featureID, proActiv$promoterId)],
         MCF7_class = proActiv$MCF7_class[match(featureID, proActiv$promoterId)],
         A549_proActiv = proActiv$A549[match(featureID, proActiv$promoterId)],
         MCF7_proActiv = proActiv$MCF7[match(featureID, proActiv$promoterId)],
         promoterCount = proActiv$promoterCount[match(featureID, proActiv$promoterId)],
         geneExpDiff = abs(geneExp$A549-geneExp$MCF7)[match(groupID, geneExp$geneId)]) %>%
  group_by(groupID) %>%
  filter(any(xor(A549_class == 'Major', MCF7_class == 'Major') & geneExpDiff < 1 & promoterCount < 5) & all(!is.na(padj)))%>%
  mutate(minP = min(padj)) %>%
  arrange(minP)
```

```{r eval=TRUE, echo=FALSE}
knitr::kable(head(diffPromoter %>% dplyr::select(-pvalue, -promoterCount, -minP)))
```

## Visualization

Here we offer several visualizations of the data returned by the worfklow above.

### Differential Promoter Usage

To visualize genes with alternate promoter usage across conditions, we use Gviz and data from `diffPromoter`. In particular, we visualize the gene corresponding to the promoter with the highest statistical significance called by DEXSeq, tumour protein 63.

```{r eval=TRUE, message=FALSE, fig.height=4.5, fig.width=7, fig.align='center'}
library(Gviz)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)

## Create DataTrack for promoter activity 
gene <- diffPromoter$groupID[1]

geneCoord <- proCoord %>% filter(geneId == gene)

geneMap <- promoterIdMapping(promoterAnnotation) %>% filter(geneId == gene)

geneRes <- diffPromoter %>% filter(groupID == gene)

gr <- GRanges(seqnames = geneCoord$seqnames[1], 
              strand = geneCoord$strand,
              ranges = IRanges(start = geneCoord$start, width = 3000))

values(gr) <- DataFrame(A549 = geneRes$A549_proActiv, 
                        MCF7 = geneRes$MCF7_proActiv)

maxy <- max(unlist(mcols(gr))) + 0.5

dtrack1 <- DataTrack(gr[,groups[1]], name = paste(groups[1], '\n Promoter Activity'), ylim = c(0, maxy))

dtrack2 <- DataTrack(gr[,groups[2]], name = paste(groups[2], '\n Promoter Activity'), ylim = c(0, maxy))
                
## Create Genome Axis Track
gtrack <- GenomeAxisTrack() 

## Create Ideogram Track
itrack <- IdeogramTrack(genome = 'hg38', chromosome = geneCoord$seqnames[1])

## Create Gene Model Track
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

txs <- as_tibble(transcripts(txdb)) %>% dplyr::filter(tx_name %in% geneMap$transcriptName)

exByTx <- exonsBy(txdb, by = 'tx')

ex_gr <- GRangesList(lapply(txs$tx_id, function(x) exByTx[[x]]))

exs <- as_tibble(sort(unique(unlist(ex_gr)))) %>% mutate(gene = gene)

gmtrack <- GeneRegionTrack(exs,
                      genome = 'hg38',
                      chr = exs$seqnames[1],
                      start = exs$start[1],
                      end = exs$end[nrow(exs)],
                      collapseTranscripts = 'meta',
                      name = 'All Exons')

plotTracks(list(itrack, dtrack1, dtrack2, gmtrack, gtrack), 
           type='histo', main = gene, 
           col.axis = 'black', col.title = 'black', background.title = 'transparent',
           cex.title = 0.7, cex.axis = 0.7, cex.main = 0.8)
```

### Promoter Category 

Analysis of the proportion of promoters falling into each of the three categories for each cell line. 

```{r eval=TRUE, message=FALSE, warning=FALSE}
library(ggplot2)

proActivLong <- tidyr::pivot_longer(proActiv, paste0(groups, '_class'), 
                             names_to = 'sample', values_to = 'promoterClass', values_drop_na = TRUE)

promoterProportions <- ggplot(subset(proActivLong, !is.na(promoterClass))) +
  geom_bar(aes(x = sample, fill = factor(promoterClass, levels = c('Inactive', 'Minor', 'Major'))), color = 'black') +
  ggtitle('Categorization of Promoters') + 
  xlab('Cell Lines') + ylab('Count') + scale_x_discrete(labels = groups) +
  labs(fill = 'Promoter Category') + 
  scale_fill_manual(labels = c('Inactive (< 0.25)', 'Minor', 'Major (Max)'), values = c('#F8766D', '#00BFC4', '#C77CFF')) +
  theme_minimal()
```

```{r echo=FALSE, fig.height=5, fig.width=5, fig.align='center'}
promoterProportions
```

### Major/Minor Promoters by Position

Analysis of major:minor promoter proportions against promoter position for chromosome 3. The analysis is restricted to multi-promoter genes with at least one active promoter. In general, the major:minor promoter proportion decreases with increasing promoter position.  

```{r eval=TRUE, message=FALSE, warning=FALSE}
proActivSub <- proActiv %>%
  group_by(geneId) %>%
  filter(any(A549_class == 'Major') & promoterCount > 1)

proPropByPos <- ggplot(subset(proActivSub, A549_class %in% c('Major', 'Minor'))) +
  geom_bar(aes(x = promoterRank, fill = factor(A549_class, levels = c('Minor', 'Major'))), position = 'fill', color = 'black') +
  ggtitle('Major/Minor Promoter Proportion in A549') + 
  xlab(expression(Promoter ~ Position ~ "5'" %->% "3'")) + ylab('Proportion of Promoters') +
  scale_y_continuous(breaks = seq(0,1, 0.25), labels = paste0(seq(0,100,25),'%')) +
  scale_x_continuous(breaks = seq(1,5), labels = c('1','2','3','4','>=5')) +
  labs(fill = 'Promoter Category') + 
  scale_fill_manual(labels = c('Minor (>= 0.25)', 'Major (Max)'), values = c('#00BFC4', '#C77CFF')) +
  theme_minimal()
```

```{r echo=FALSE, fig.height=5, fig.width=5, fig.align='center'}
proPropByPos
```

 
### Major Promoter Activity and Gene Expression

Comparison of major promoter activity and gene expression, calculated by summing over all promoters. Single promoter genes lie on the diagonal. Multi-promoter genes lie to the right of the diagonal. A single major promoter does not often fully explain gene expression, with minor promoters also contributing to gene expression.

```{r eval=TRUE, message=FALSE, warning=FALSE}

majorA549 <- proActiv %>% filter(A549_class == 'Major')

geneExpA549 <- geneExp %>% filter(geneId %in% majorA549$geneId)

majorA549$geneExp <- geneExpA549$A549

geneExpMajorPro <- ggplot(majorA549, aes(x = geneExp, y = A549)) + 
  geom_point(aes(colour = promoterCount), shape = 1) +
  ggtitle('Major Promoter Activity vs. Gene Expression in A549') + 
  xlab('Average Gene Expression') + ylab('Average Major Promoter Activity') +
  labs(colour = 'Number of \n Active Promoters') +
  geom_abline(slope = 1, intercept = 0, colour = 'red', linetype = 'dashed') +
  theme_minimal() 
```

```{r echo=FALSE, fig.height=5, fig.width=6, fig.align='center'}
geneExpMajorPro
```


### t-SNE 

We generate a t-SNE plot using the active promoters from chromosome 3. Expectedly, replicates from each cell line cluster together. 

```{r, eval=TRUE}
library(Rtsne)

data <- assays(result)$absolutePromoterActivity %>% 
  na.omit %>% 
  filter(rowSums(.) > 0)

data <- data.frame(t(data))

data$Sample <- as.factor(unlist(lapply(strsplit(rownames(data), '_'), `[[`, 2)))

set.seed(29) ## for reproducibility

tsne.out <- Rtsne(as.matrix(subset(data, select = -c(Sample))), perplexity = 1)
```


```{r fig.height=5, fig.width=5.2, fig.align='center'}
plot(x = tsne.out$Y[,1], y = tsne.out$Y[,2], bg = data$Sample, asp = 1,
     col = 'black', pch = 24, cex = 4,
     main = 't-SNE plot with promoters \n active in at least one sample',
     xlab = 'T-SNE1', ylab = 'T-SNE2', 
     xlim = c(-300,300), ylim = c(-300,300))

legend('topright', inset = .02, title = 'Cell Lines',
       groups, pch = c(24,24), pt.bg = 1:length(groups) , cex = 1.5, bty = 'n')

```

## Reference

If you use proActiv, please cite:

[DemircioÄŸlu, Deniz, et al. "A Pan-cancer Transcriptome Analysis Reveals Pervasive Regulation through Alternative Promoters." *Cell* 178.6 (2019): 1465-1477.](https://www.cell.com/cell/fulltext/S0092-8674(19)30906-7)


