<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Estimate Promoter Activity from RNA-Seq data • proActiv</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Estimate Promoter Activity from RNA-Seq data">
<meta property="og:description" content="Estimate promoter activity using RNA-Seq data. 
  This package uses BAM files or junction files from Tophat2
  or STAR and estimates promoter activity. The promoters of each 
  gene can be identified by providing a txdb object for the input 
  annotation version. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">proActiv</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="articles/proActiv.html">
    <span class="fa fa-file-alt"></span>
     
    Complete Analysis Workflow
  </a>
</li>
<li>
  <a href="reference/index.html">
    <span class="fa fa-code"></span>
     
    Manual
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GoekeLab/proActiv/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="reference/figures/proActiv_design.png" alt="Stay proActiv!"></p>
<p><img src="reference/figures/proActiv_name.png" alt="Stay proActiv!"></p>
<div id="proactiv-estimation-of-promoter-activity-from-rna-seq-data" class="section level2">
<h2 class="hasAnchor">
<a href="#proactiv-estimation-of-promoter-activity-from-rna-seq-data" class="anchor"></a>proActiv: Estimation of Promoter Activity from RNA-Seq data</h2>
<!-- badges: start -->
<p><a href="https://github.com/GoekeLab/proActiv/releases/"><img src="https://img.shields.io/github/v/release/GoekeLab/proActiv" alt="GitHub release (latest by date)"></a> <a href="https://github.com/GoekeLab/proActiv/graphs/contributors"><img src="https://img.shields.io/badge/Maintained%3F-Yes-brightgreen" alt="Maintained?"></a> <a href="#installation"><img src="https://img.shields.io/badge/Install-Github-brightgreen" alt="Install"></a> <!-- badges: end --></p>
<p>proActiv is an R package that estimates promoter activity from RNA-Seq data. proActiv uses aligned reads and genome annotations as input, and provides absolute and relative promoter activity as output. The package can be used to identify active promoters and alternative promoters. Details of the method are described in <a href="#reference">Demircioglu et al</a>.</p>
<p>HTML documentation of proActiv, including a complete step-by-step workflow and a function manual, is available at <a href="https://goekelab.github.io/proActiv/" class="uri">https://goekelab.github.io/proActiv/</a>.</p>
<p>Additional data on differential promoters in tissues and cancers from TCGA, ICGC, GTEx, and PCAWG is available at <a href="https://jglab.org/data-and-software/" class="uri">https://jglab.org/data-and-software/</a>.</p>
<div id="content" class="section level3">
<h3 class="hasAnchor">
<a href="#content" class="anchor"></a>Content</h3>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#quick-start">Quick Start</a></li>
<li><a href="#creating-a-promoter-annotation-object">Creating a Promoter Annotation object</a></li>
<li><a href="#complete-analysis-workflow-analyzing-alternative-promoters">Complete Analysis Workflow: Analyzing Alternative Promoters</a></li>
<li><a href="#limitations">Limitations</a></li>
<li><a href="#release-history">Release History</a></li>
<li><a href="#reference">Reference</a></li>
<li><a href="#contributors">Contributors</a></li>
</ul>
</div>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<p>proActiv can be installed from GitHub with:</p>
<div class="sourceCode" id="cb1"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"devtools"</span>)
<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"GoekeLab/proActiv"</span>)</pre></div>
</div>
<div id="quick-start" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h3>
<p>proActiv estimates promoter activity from RNA-Seq data. Promoter activity is defined as the total amount of transcription initiated at each promoter. proActiv takes as input either BAM files or junction files (TopHat2 or STAR), and a promoter annotation object of the relevant genome. An optional argument <code>condition</code> can be supplied, describing the experimental condition corresponding to each input file. Here we demonstrate proActiv with STAR junction files (Human genome GRCh38 GENCODE v34) as input. These files are taken from the <a href="https://github.com/GoekeLab/sg-nex-data">SGNEx project</a>, and can be found at ‘extdata/vignette’:</p>
<ul>
<li>extdata/vignette/SGNEx_A549_Illumina_replicate1-run1.junctions.gz</li>
<li>extdata/vignette/SGNEx_A549_Illumina_replicate3-run1.junctions.gz</li>
<li>extdata/vignette/SGNEx_A549_Illumina_replicate5-run1.junctions.gz</li>
<li>extdata/vignette/SGNEx_HepG2_Illumina_replicate2-run1.junctions.gz</li>
<li>extdata/vignette/SGNEx_HepG2_Illumina_replicate4-run1.junctions.gz</li>
<li>extdata/vignette/SGNEx_HepG2_Illumina_replicate5-run1.junctions.gz</li>
</ul>
<!-- end list --><div class="sourceCode" id="cb2"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">proActiv</span>)

<span class="co">## List of STAR junction files as input</span>
<span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/vignette/junctions'</span>,
                                <span class="kw">package</span> <span class="kw">=</span> <span class="st">'proActiv'</span>), <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">## Vector describing experimental condition</span>
<span class="no">condition</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'A549'</span>,<span class="st">'HepG2'</span>), <span class="kw">each</span><span class="kw">=</span><span class="fl">3</span>)
<span class="co">## Promoter annotation for human genome GENCODE v34</span>
<span class="no">promoterAnnotation</span> <span class="kw">&lt;-</span> <span class="no">promoterAnnotation.gencode.v34</span>

<span class="no">result</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/proActiv.html">proActiv</a></span>(<span class="kw">files</span> <span class="kw">=</span> <span class="no">files</span>,
                   <span class="kw">promoterAnnotation</span> <span class="kw">=</span> <span class="no">promoterAnnotation</span>,
                   <span class="kw">condition</span> <span class="kw">=</span> <span class="no">condition</span>)</pre></div>
<p><code>result</code> is a summarizedExperiment object which can be accessed as follows:</p>
<ul>
<li>
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assays(results)</a></code> returns raw/normalized promoter counts and absolute/relative promoter activity<br>
</li>
<li>
<code><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata(results)</a></code> returns gene expression data<br>
</li>
<li>
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData(results)</a></code> returns promoter metadata and summarized absolute promoter activity by conditions</li>
</ul>
<p>proActiv can also be run with BAM files as input, but an additional parameter <code>genome</code> must be supplied:</p>
<div class="sourceCode" id="cb3"><pre class="r"><span class="co">## From BAM files - genome parameter must be provided</span>
<span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/testdata/bam'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'proActiv'</span>), <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">result</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/proActiv.html">proActiv</a></span>(<span class="kw">files</span> <span class="kw">=</span> <span class="no">files</span>,
                   <span class="kw">promoterAnnotation</span> <span class="kw">=</span> <span class="no">promoterAnnotation.gencode.v34</span>,
                   <span class="kw">genome</span> <span class="kw">=</span> <span class="st">'hg38'</span>)</pre></div>
</div>
<div id="creating-a-promoter-annotation-object" class="section level3">
<h3 class="hasAnchor">
<a href="#creating-a-promoter-annotation-object" class="anchor"></a>Creating a Promoter Annotation object</h3>
<p>In order to quantify promoter activity, proActiv uses a set of promoters based on genome annotations. proActiv allows the creation of a promoter annotation object for any genome from a TxDb object or from a GTF file with the <code>preparePromoterAnnotation</code> function. Users have the option to either pass the file path of the GTF/GFF or TxDb to be used, or use the TxDb object directly as input. proActiv includes pre-calculated promoter annotations for human and mouse genomes:</p>
<ul>
<li>GENCODE Release 19 / GRCh37 / hg19 : <code>promoterAnnotation.gencode.v19</code>
</li>
<li>GENCODE Release 34 / GRCh38 / hg38 : <code>promoterAnnotation.gencode.v34</code>
</li>
<li>GENCODE Release M1 / NCBIM37 / mm9 : <code>promoterAnnotation.gencode.vM1</code>
</li>
<li>GENCODE Release M25 / GRCm38.p6 / mm10 : <code>promoterAnnotation.gencode.vM25</code>
</li>
</ul>
<p>GTF files can be downloaded from the <a href="https://www.gencodegenes.org">GENCODE</a> page.</p>
<p>Here, we demonstrate creating the promoter annotation for the Human genome (GENCODE v34) with both GTF and TxDb. To keep the run-time small, we use a subset of the GTF/TxDb which includes only chromosome 22 annotations.</p>
<div class="sourceCode" id="cb4"><pre class="r"><span class="co">## From GTF file path</span>
<span class="no">gtf.file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/vignette/annotation/gencode.v34.annotation.chr22.gtf.gz'</span>,
                        <span class="kw">package</span> <span class="kw">=</span> <span class="st">'proActiv'</span>)
<span class="no">promoterAnnotation.gencode.v34.chr22</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/preparePromoterAnnotation.html">preparePromoterAnnotation</a></span>(<span class="kw">file</span> <span class="kw">=</span> <span class="no">gtf.file</span>,
                                                                  <span class="kw">species</span> <span class="kw">=</span> <span class="st">'Homo_sapiens'</span>)
<span class="co">## From TxDb object</span>
<span class="no">txdb.file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/vignette/annotation/gencode.v34.annotation.chr22.sqlite'</span>,
                         <span class="kw">package</span> <span class="kw">=</span> <span class="st">'proActiv'</span>)
<span class="no">txdb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">loadDb</a></span>(<span class="no">txdb.file</span>)
<span class="no">promoterAnnotation.gencode.v34.chr22</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/preparePromoterAnnotation.html">preparePromoterAnnotation</a></span>(<span class="kw">txdb</span> <span class="kw">=</span> <span class="no">txdb</span>,
                                                                  <span class="kw">species</span> <span class="kw">=</span> <span class="st">'Homo_sapiens'</span>)</pre></div>
<p>The <code>PromoterAnnotation</code> object has 3 slots:</p>
<ul>
<li>
<code>intronRanges</code> : Intron ranges, giving the corresponding transcripts of each intron</li>
<li>
<code>promoterIdMapping</code> : An ID mapping between transcripts, promoter IDs and gene IDs<br>
</li>
<li>
<code>promoterCoordinates</code> : Promoter coordinates (TSS) and internal promoter state, along with the 3’ coordinate of the first exon</li>
</ul>
</div>
<div id="complete-analysis-workflow-analyzing-alternative-promoters" class="section level3">
<h3 class="hasAnchor">
<a href="#complete-analysis-workflow-analyzing-alternative-promoters" class="anchor"></a>Complete Analysis Workflow: Analyzing Alternative Promoters</h3>
<p>Most human genes have multiple promoters that control the expression of distinct isoforms. The use of these alternative promoters enables the regulation of isoform expression pre-transcriptionally. Importantly, alternative promoters have been found to be important in a wide number of cell types and diseases. proActiv includes a workflow to identify and visualize alternative promoter usage between conditions. This workflow is described in detail <a href="https://goekelab.github.io/proActiv/articles/proActiv.html">here</a>.</p>
</div>
</div>
<div id="release-history" class="section level2">
<h2 class="hasAnchor">
<a href="#release-history" class="anchor"></a>Release History</h2>
<p><strong>Release 0.99.0</strong></p>
<p>Release date: 4th August 2020</p>
<p>Changes in version 0.99.0:</p>
<ul>
<li><p>Workflow: The wrapper function <code>proActiv</code> performs all steps to estimate promoter activity and calculates promoter metadata. A <code>condition</code> argument can be supplied for <code>proActiv</code> to summarize promoter counts and activity across conditions. These results are returned as a SummarizedExperiment object.</p></li>
<li><p>BAM file usage: In addition to junction files, <code>proActiv</code> now allows BAM files as input. However, users should note that this function is not fully optimized and may have long run-time.</p></li>
<li><p>Promoter annotation: Improved efficiency in generating promoter annotations without the need for parallelization with the <code>preparePromoterAnnotations</code> function. Promoter annotation objects for human (hg19/hg38) and mouse (mm9/mm10) genomes are now pre-calculated and available to the user. The promoter annotation object is also trimmed to preserve essential information for running <code>proActiv</code>, in order to comply with Bioconductor guidelines concerning package size.</p></li>
<li><p>Plotting promoter activity: The plotting function <code>plotPromoters</code> visualizes promoter activity across conditions. It accepts the SummarizedExperiment object returned by <code>proActiv</code> along with a gene of interest and gene annotations as arguments. This allows users to visualize promoter activity and identify instances of alternative promoter usage.</p></li>
<li><p>Vignette: proActiv now comes with a vignette, documenting a complete step-by-step workflow in identifying active and alternative promoter usage. This includes guidance on running <code>proActiv</code>, creating promoter annotations and identifying alternative promoter usage. Various visualizations of promoter activity are also offered.</p></li>
</ul>
<p><strong>Initial Release 0.1.0</strong></p>
<p>Release date: 19th May 2020</p>
<p>This release corresponds to the proActiv version used by <a href="#reference">Demircioglu et al.</a></p>
</div>
<div id="limitations" class="section level2">
<h2 class="hasAnchor">
<a href="#limitations" class="anchor"></a>Limitations</h2>
<p>proActiv will not provide promoter activity estimates for promoters which are not uniquely identifiable from splice junctions (single exon transcripts, promoters which overlap with internal exons).</p>
</div>
<div id="reference" class="section level2">
<h2 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h2>
<p>If you use proActiv, please cite:</p>
<p><a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30906-7">Demircioğlu, Deniz, et al. “A Pan-cancer Transcriptome Analysis Reveals Pervasive Regulation through Alternative Promoters.” <em>Cell</em> 178.6 (2019): 1465-1477.</a></p>
</div>
<div id="contributors" class="section level2">
<h2 class="hasAnchor">
<a href="#contributors" class="anchor"></a>Contributors</h2>
<p>proActiv is developed and maintained by <a href="https://github.com/dnzdmrcgl">Deniz Demircioglu</a>, <a href="https://github.com/jleechung">Joseph Lee</a>, and <a href="https://github.com/jonathangoeke">Jonathan Göke</a>.</p>
<p><img src="reference/figures/proActiv_logoName.png" alt="Stay proActiv!"></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/GoekeLab/proActiv/">https://​github.com/​GoekeLab/​proActiv/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing proActiv</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Deniz Demircioglu <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-7857-0407" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jonathan Göke <br><small class="roles"> Author </small>  </li>
<li>Joseph Lee <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Deniz Demircioglu, Jonathan Göke, Joseph Lee.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
